zerberus/sfz.cpp

```diff
@@ -465,76 +465,146 @@ void SfzRegion::readOp(const QString& b, const QString& data, SfzControl &c)
             qDebug("SfzRegion: unknown opcode <%s>", qPrintable(opcode));
       }
 
+QStringList readFile(QString const fn)
+      {
+            QFile f(fn);
+            QStringList list;
+            if (!f.open(QIODevice::ReadOnly)) {
+                  qDebug("ZInstrument: cannot load %s", qPrintable(fn));
+                  return list;
+                  }
+            while (!f.atEnd())
+                  list.append(QString(f.readLine().simplified()));
+            return list;
+      }
+
 //---------------------------------------------------------
 //   loadSfz
 //---------------------------------------------------------
 
 bool ZInstrument::loadSfz(const QString& s)
       {
       _program = 0;
-      QFile f(s);
-      if (!f.open(QIODevice::ReadOnly)) {
-            qDebug("ZInstrument: cannot load %s", qPrintable(s));
+      QFileInfo fi(s);
+      QString path = fi.absolutePath();
+
+      QStringList fileContents = readFile(s);
+
+      if (fileContents.empty()) {
             return false;
             }
-      QFileInfo fi(f);
-      QString path = fi.absolutePath();
-      qint64 total = fi.size();
 
-      QString sample;
+      SfzControl c;
+      c.init();
+      for (int i = 0;i < 128; i++)
+            c.set_cc[i] = -1;
+
+      int idx = 0;
+      // preprocessor
+      while(idx < fileContents.size()) {
+            QRegularExpression findWithSpaces("\"(.+)\"");
+            QRegularExpressionMatch foundWithSpaces;
+            QString curLine = fileContents[idx];
+
+            if (curLine.startsWith("#define")) {
+                  QStringList define = curLine.split(" ");
+                  foundWithSpaces = findWithSpaces.match(curLine);
+                  if (define.size() == 3)
+                        c.defines.insert(std::pair<QString, QString>(define[1], define[2]));
+                  else if(foundWithSpaces.hasMatch())
+                        c.defines.insert(std::pair<QString, QString>(define[1], foundWithSpaces.captured(1)));
+                  fileContents.removeAt(idx);
+                  }
+            else if (curLine.startsWith("#include")) {
```

Comment:
They should work. As you can see the contents of the file are inserted after the current line -> so if they contain another #include they should be inserted there etc.